Adding TRR for SAM Dumping (T1003.002) (#22)
tired-labs/techniques: reports/trr0000/win/README.md

│   │   │   ├── 000001F5\         ← Guest (RID 501)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── 000001F7\         ← DefaultAccount (RID 503)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── 000003E8\         ← First user-created account (RID 1000)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── Names\            ← maps usernames to RIDs
│   │   │   │   ├── Administrator
│   │   │   │   ├── Guest
│   │   │   │   └── ...
│   │   └── Aliases\              ← local groups
│   └── Builtin\                  ← built-in groups
```

Each local user account has an entry under `SAM\Domains\Account\Users\[RID]`,
where `[RID]` is the account's relative identifier (e.g., `000001F4` for the
built-in Administrator). Each key has the following values:

- `V` value - contains the encrypted hashes and profile data
- `F` value - contains account operational data: last logon time, password last
  set, account expiration, login count, failed login count, account control
  flags (disabled, locked out, etc.)
- `SupplementalCredentials` - contains a structured binary value that holds

>   additional cryptographic forms of the user's password.

Reviewer: The SAM Database section introduces three values under each user's RID key: `V`, `F`, and `SupplementalCredentials`. The description of `SupplementalCredentials` states it:

> contains a structured binary value that holds additional cryptographic forms of the user's password.

However, none of the procedures reference `SupplementalCredentials` during extraction. Procedure A's enumerate-and-query approach only describes reading the `V` value, and the decryption step doesn't mention it either.

Could you clarify the relationship between `SupplementalCredentials` and the procedures? If it is relevant to credential extraction, should the enumerate-and-query approach include it as a value to extract? If it is not, could the Technical Background note why so a reader doesn't expect to see it addressed later in the document?

Author: Correct, this key is not used in this technique. I've added a footnote that provides more details on what the key does hold and a note that it is not required for any procedure in this technique.

operation with no direct telemetry.

> [!Note]
>
> In addition to the direct API calls, [PowerShell's registry drives], the
> `RegEdit.exe` application, and the `reg.exe` utility use the standard Windows
> registry APIs and thus also fall under this procedure.

For both approaches, the user needs to be part of a group that can enable
`SeBackupPrivilege` on the target machine and all registry API calls must have
the `REG_OPTION_BACKUP_RESTORE` flag set.

#### Detection Data Model

![DDM - Local Registry Access](ddms/trr0000_win_a.png)

The DDM consists of three steps:

1. Extract key data from the `SYSTEM` hive
2. Extract hashes and key material from the `SAM` hive
3. Decrypt the hashes

Steps 1 and 2 can be done in any order but must both be completed before step
3. Because of the complexity of the first two steps, the DDM contains a
high-level initial model, with 'implementation' arrows showing the specific
sub-model for each high-level step.

For both steps 1 and 2, the model shows the **enumerate-and-query** approach and
the **save** approach. Both have been included in this procedure because an
attacker can mix and match them as they choose.

> ### Procedure B: Remote Registry Access

Reviewer: The TRR states that the Remote Registry service is:

> disabled by default on standard Windows 10 and 11 machines

and

> must be enabled manually or via group policy

This is accurate for workstations but does not cover Windows Server behavior, which is where I believe it is most likely to be used in practice. The service's default state varies by platform:

- Windows Server 2003 and Windows XP/2000: The service is configured to start automatically by default. (Remarks section @ [RegConnectRegistryA - Microsoft Learn](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regconnectregistrya))
- Windows Server 2012+: You wanna hear something that blew my mind?! The service is set to `Manual (Trigger Start)`. It auto-starts when a connection is made to the `winreg` named pipe, meaning no administrator intervention is required for Procedure B to succeed. ([robertwray.co.uk - Manual Trigger Start](https://robertwray.co.uk/blog/what-does-manual-triggered-mean-for-a-windows-service))
- Windows Server 2019+: This one might be worth adding just for helping emulation tests not cause the authors to pull their hair out. The service stops after 10 minutes of idle time by design. ([Microsoft Q&A](https://learn.microsoft.com/en-us/answers/questions/126835/remote-register-service-keep-stop))
- Windows 10/11: Disabled by default (as the TRR currently states).

Consider expanding the prerequisite discussion for Procedure B to cover these platform-specific behaviors.

Author: On Server 2019+, if I'm reading it correctly, the service will shut down after 10 mins, but it will still be auto-restarted (trigger start) if a new request is received. https://blog.compass-security.com/2022/05/bloodhound-inner-workings-part-3/.  So behavior should feel the same for everything Server 2012+.

Author: Updated, lemme know what you think.

Reviewer: All of my comments around this are theoretical as I haven't proved them out in my home lab so I trust your instincts. Looks good to me!

access events are not generated. Some EDR products have visbility into volume
handle operations and can observe `CreateFile` calls targeting raw disks or
volumes.

### Procedure E: Acquire Registry Backup

Certain versions of Windows automatically back up critical registry files,
include the SAM and SYSTEM hives. Starting with Windows 10, Microsoft disabled
these automatic backups on personal desktop versions of the operating system
(Windows 10 and 11).[^4] It remains on by default for server versions. If
disabled, this behavior can be re-enabled by creating a DWORD registry value
named `EnablePeriodicBackup` set to `1` under
`HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Configuration Manager`.

Once set, the `RegIdleBackup` scheduled task will create backup copies of all
registry hives (including SAM and SYSTEM) during idle maintenance periods once
every 10 days after a system restart. The `RegIdleBackup` scheduled task has a
COM handler with `ClassId` `{CA767AA8-9157-4604-B64B-40747123D5F2}`, which is
handled by `Windows\System32\regidle.dll`. This library makes a call to the
undocumented `NtInitializeRegistry` function with an argument of `0x13e8`, which

> is implemented in `ntoskrnl32.exe`. The kernel code does a number of checks,

Reviewer: ```suggestion
is implemented in `ntoskrnl.exe`. The kernel code does a number of checks,
```

│   │   │   │   ├── V             ← encrypted user data (hashes, username, etc.)
|   |   |   |   └── F             ← account operational data (last logon, logon count, etc.)
│   │   │   ├── 000001F5\         ← Guest (RID 501)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── 000001F7\         ← DefaultAccount (RID 503)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── 000003E8\         ← First user-created account (RID 1000)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── Names\            ← maps usernames to RIDs
│   │   │   │   ├── Administrator
│   │   │   │   ├── Guest
│   │   │   │   └── ...
│   │   └── Aliases\              ← local groups
│   └── Builtin\                  ← built-in groups
```

Each local user account has an entry under `SAM\Domains\Account\Users\[RID]`,
where `[RID]` is the account's relative identifier (e.g., `000001F4` for the
built-in Administrator). Each key has the following values:

- `V` value - contains the encrypted hashes and profile data
- `F` value - contains account operational data: last logon time, password last
  set, account expiration, login count, failed login count, account control

>   flags (disabled, locked out, etc.)
> - `SupplementalCredentials` - contains a structured binary value that holds
>   additional forms are used in this technique.

Author: ```suggestion
- `SupplementalCredentials` - contains a structured binary value that holds
  additional cryptographic forms of the user's password.[^10] None of these
  additional forms are used in this technique.
```

│   │   │   ├── 000001F5\         ← Guest (RID 501)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── 000001F7\         ← DefaultAccount (RID 503)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── 000003E8\         ← First user-created account (RID 1000)
│   │   │   │   └── V
|   |   |   |   └── F 
│   │   │   ├── Names\            ← maps usernames to RIDs
│   │   │   │   ├── Administrator
│   │   │   │   ├── Guest
│   │   │   │   └── ...
│   │   └── Aliases\              ← local groups
│   └── Builtin\                  ← built-in groups
```

Each local user account has an entry under `SAM\Domains\Account\Users\[RID]`,
where `[RID]` is the account's relative identifier (e.g., `000001F4` for the
built-in Administrator). Each key has the following values:

- `V` value - contains the encrypted hashes and profile data
- `F` value - contains account operational data: last logon time, password last
  set, account expiration, login count, failed login count, account control
  flags (disabled, locked out, etc.)
- `SupplementalCredentials` - contains a structured binary value that holds
  additional cryptographic forms of the user's password.[^10] None of these

>   additional forms are not used in this technique.

Reviewer: > None of these ... are not used ...

This is worded confusingly. Maybe remove the double negative? After this, lgtm!

Author: ```suggestion
  additional forms are used in this technique.
```

*Laughs*  Yup, nothing confusing at all about "none of these are not used."  That's up there with "all of your bases are belong to us."

- [CVE-2021-36934 - MSRC]
- [Silent Harvest: Extracting Windows Secrets Under the Radar - Sud0ru blog]
- [Mimikatz lsadump Module - wiki.yourway]
- [regsecrets.py - GitHub]
- [Revisiting SecretsDump - Synacktiv]
- [Priv2Admin: SeBackupPrivilege - gtworek]
- [Atomic Test T1003.002]

[^1]: `OpenLocalMachine` is an initially confusing name for a procedure that is
    connecting to a remote system's registry. The name refers to the fact that
    the procedure is used to open the `HKEY_LOCAL_MACHINE` hive on the remote
    system. There is an `Open*` procedure for each hive; for example,
    `OpenClassesRoot` will open `HKEY_CLASSES_ROOT` on the remote system.
[^2]: [BitLocker I/O Control - Geoff Chappell]
[^3]: [Unlocking BitLocker - hackyboiz]
[^4]: [No Backed Up RegBack Folder - Microsoft Learn]
[^5]: [Volume Shadow Copy API - Microsoft Learn]
[^6]: [Code Example: The Open Dialog Box - Microsoft Learn]
[^7]: [VSS: Generating a Backup Set - Microsoft Learn]
[^8]: [VShadow - GitHub]
[^9]: [An NTFS Parser Library - CodeProject]
[^10]: [MS-SAMR SupplementalCredentials - Microsoft Learn]
[^11]: See the footnotes at [RegConnectRegistryA function - Microsoft Learn]

> [^12]: BloodHound Inner Workings - Compass Security]

Author: ```suggestion
[^12]: [BloodHound Inner Workings - Compass Security]
```
